tests/integration/testvalidator.nim

```diff
@@ -0,0 +1,313 @@
+      inc i
+      secondsElapsed = i*step
+      # echo secondsElapsed.seconds
+      if secondsElapsed mod 180 == 0:
+        await stopTrackingEvents()
+        await marketplace.startTrackingEvents()
+      if endTime < Moment.now():
+        return false
+      if cancelExpression:
+        return false
+      await sleepAsync(step.seconds)
+    return true
+
+  await eventuallyS()
+
+marketplacesuiteWithProviderUrl "Validation", "ws://localhost:8545":
+  let nodes = 3
+  let tolerance = 1
+  let proofProbability = 1
+
+  var events = {
+    $SlotFilled: newSeq[ref MarketplaceEvent](),
+    $SlotFreed: newSeq[ref MarketplaceEvent](),
+    $RequestFailed: newSeq[ref MarketplaceEvent](),
+    $RequestCancelled: newSeq[ref MarketplaceEvent]()
+  }.toTable
+  var eventSubscriptions = newSeq[provider.Subscription]()
+
+  proc box[T](x: T): ref T =
+    new(result);
+    result[] = x
+
+  proc onMarketplaceEvent[T: MarketplaceEvent](event: T) {.gcsafe, raises:[].} =
+    try:
+      debug "onMarketplaceEvent", eventType = $T, event = event
+      events[$T].add(box(event))
+    except KeyError:
+      discard
+
+  proc startTrackingEvents(marketplace: Marketplace) {.async.} =
+    eventSubscriptions.add(
+      await marketplace.subscribe(SlotFilled, onMarketplaceEvent[SlotFilled])
+    )
+    eventSubscriptions.add(
+      await marketplace.subscribe(RequestFailed, onMarketplaceEvent[RequestFailed])
+    )
+    eventSubscriptions.add(
+      await marketplace.subscribe(SlotFreed, onMarketplaceEvent[SlotFreed])
+    )
+    eventSubscriptions.add(
+      await marketplace.subscribe(RequestCancelled, onMarketplaceEvent[RequestCancelled])
+    )
+  
+  proc stopTrackingEvents() {.async.} =
+    for event in eventSubscriptions:
+      await event.unsubscribe()
+  
+  proc checkSlotsFreed(requestId: RequestId, expectedSlotsFreed: int): bool =
+    events[$SlotFreed].filter(
+      e => (ref SlotFreed)(e).requestId == requestId)
+        .len == expectedSlotsFreed and
+      events[$RequestFailed].map(
+        e => (ref RequestFailed)(e).requestId).contains(requestId)
+  
+  proc isRequestCancelled(requestId: RequestId): bool =
+    events[$RequestCancelled].map(e => (ref RequestCancelled)(e).requestId)
+      .contains(requestId)
+  
+  proc getSlots[T: MarketplaceEvent](requestId: RequestId): seq[SlotId] =
+    events[$T].filter(
+      e => (ref T)(e).requestId == requestId).map(
+        e => slotId((ref T)(e).requestId, (ref T)(e).slotIndex))
+  
+  proc checkSlotsFailed(marketplace: Marketplace, requestId: RequestId) {.async.} =
+    let slotsFreed = getSlots[SlotFreed](requestId)
+    let slotsFilled = getSlots[SlotFilled](requestId)
+    let slotsNotFreed = slotsFilled.filter(
+      slotId => not slotsFreed.contains(slotId)
+    ).toHashSet
+    var slotsFailed = initHashSet[SlotId]()
+    for slotId in slotsFilled:
+      let state = await marketplace.slotState(slotId)
+      if state == SlotState.Failed:
+        slotsFailed.incl(slotId)
```

Comment:
Because of erasure coding, after (tolerance + 1) slots are freed, the request will fail and the
remaining slots will be marked as "Failed". When the request has failed, `slotState` returns `Failed`
regardless of the slot being freed earlier or not.

Thus, after request has failed, for all the slots belonging to that request, the `slotState` will be `Failed`. 
But the request could have failed for other reasons, so I prefer not to solely depend on the request logic when I am testing validators.

I've also provided some extra info [here](https://github.com/codex-storage/nim-codex/pull/922#discussion_r1804668418) - it was like a month ago - maybe you missed it.

Reply:
> But the request could have failed for other reasons

How else can the request fail?

> so I prefer not to solely depend on the request logic when I am testing validators

If you can't rely on the `codex-contracts-eth` tests and formal verification for the "request logic", then you can't rely on them for guaranteed events emission either. I'm not saying to *not* trust it, I'm saying we *should* trust it.

Reply:
Using `await marketplace.requestState(requestId)` as described in https://github.com/codex-storage/nim-codex/pull/922#discussion_r1870519963